// @aiot-toolkit/parser/lib/ux/translate/vela/ScriptToTypescript.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _sharedUtils = require("@aiot-toolkit/shared-utils");
var babel = _interopRequireWildcard(require("@babel/core"));
var _path = _interopRequireDefault(require("path"));
var _ParserUtil = _interopRequireDefault(require("../../../utils/ParserUtil"));
var _FeatureConfig = _interopRequireDefault(require("../../config/FeatureConfig"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
const BabelPreset = require('@babel/preset-env');

/**
 * ScriptToTypescript
 */
class ScriptToTypescript {
  _systemFeatures = (() => new Set())();
  constructor(options, compilerOption, context) {
    this.options = options;
    this.compilerOption = compilerOption;
    this.context = context;
  }
  translate(sourceTree,
  // eslint-disable-next-line
  offsetList) {
    // 使用babel实现替换，由此实现自动更新source map
    const content = sourceTree.content ? this.babelScript(sourceTree.content) : `""`;
    const fileName = this.options.filePath + '.js';
    const targetTree = _ParserUtil.default.createSourceFile(content, fileName);
    this.cacheAndCheckFeature();
    return {
      targetTree,
      mapList: []
    };
  }
  /**
   * babel自定义插件实现替换
   * @param sourceCode
   * @returns
   */
  babelScript(sourceCode) {
    const relativePath = _path.default.relative(this.options.projectPath, this.options.filePath);
    const plugins = [
    // 替换所有的require节点
    [_path.default.resolve(__dirname, './plugins/require.js'), {
      options: this.options,
      systemFeatures: this._systemFeatures,
      compilerOption: this.compilerOption
    }]];
    if (this.compilerOption.enableE2e) {
      plugins.push([_path.default.resolve(__dirname, './plugins/e2e.js'), {
        options: this.options,
        compilerOption: this.compilerOption
      }]);
    }
    const babelResult = babel.transformSync(sourceCode, {
      sourceMaps: 'inline',
      filename: relativePath,
      sourceFileName: relativePath,
      babelrc: false,
      presets: [[BabelPreset, {
        targets: {
          node: 8
        },
        modules: 'commonjs'
      }]],
      plugins
    });
    return babelResult?.code || '';
  }

  /**
   * 缓存并检查 feature
   */
  cacheAndCheckFeature() {
    const {
      onLog
    } = this.options;
    const errorFeatures = [];
    const translateCache = this.context.translateCache;
    this._systemFeatures.forEach(item => {
      if (!_FeatureConfig.default.isEffectSystemModule(item)) {
        errorFeatures.push(item);
      }
      if (translateCache) {
        translateCache.featureCache.add(item);
      }
    });
    if (errorFeatures.length) {
      onLog({
        filePath: this.options.filePath,
        level: _sharedUtils.Loglevel.WARN,
        message: [`unknown feature：\r\n`, {
          word: errorFeatures.map((item, index) => `${index + 1}. ${item}`).join('\r\n')
        }, '\r\n']
      });
    }
  }
}
var _default = exports.default = ScriptToTypescript;
