// @aiot-toolkit/aiotpack/lib/config/XtsConfig.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _entryTemplate = require("../afterCompile/xts/entryTemplate");
var _generateRpk = require("../afterCompile/xts/generateRpk");
var _ts2wasm = require("../afterCompile/xts/ts2wasm");
var _XtsLoader = _interopRequireDefault(require("../loader/xts/XtsLoader"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
/**
 * XtsConfig
 */
class XtsConfig {
  constructor(projectPath) {
    this.projectPath = projectPath;
  }
  exclude = ['**/node_modules{,/**}', '**/dist{,/**}', '**/build{,/**}', '**/.git{,/**}', '**/.wasmUnpacked{,/**}'];
  output = 'build';
  module = (() => ({
    rules: [{
      test: [/\.xts$/],
      loader: [_XtsLoader.default]
    }]
  }))();
  beforeCompile = [];
  afterCompile = (() => [{
    worker: _entryTemplate.generateEntryFile
  }, {
    worker: _ts2wasm.ts2wasm
  }, {
    worker: _generateRpk.generateRpk
  }])();
}
var _default = exports.default = XtsConfig;
