// @aiot-toolkit/aiotpack/lib/afterCompile/xts/ts2wasm.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.wasmPackageName = exports.ts2wasm = void 0;
var _sharedUtils = require("@aiot-toolkit/shared-utils");
var _child_process = require("child_process");
var _fastGlob = _interopRequireDefault(require("fast-glob"));
var _fs = _interopRequireDefault(require("fs"));
var _path = _interopRequireDefault(require("path"));
var _entryTemplate = require("./entryTemplate");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
const wasmPackageName = exports.wasmPackageName = 'wasmUnpacked';

/**
 * 将 __entry__.ts 文件打包成 warm 文件
 * @param context
 */
const ts2wasm = async params => {
  const {
    context,
    compilerOption
  } = params;
  /** 1. 生成 app.wasm 文件 */
  const {
    projectPath,
    output
  } = context;
  const buildDir = _path.default.resolve(projectPath, output);
  const entryFile = _path.default.resolve(buildDir, _entryTemplate.XtsEntryFileName);
  const wasmDir = _path.default.resolve(projectPath, wasmPackageName);
  _fs.default.mkdirSync(wasmDir, {
    recursive: true
  });
  function compile() {
    // TODO: 后续修改为 @mi
    const ts2wasmpath = _path.default.resolve(projectPath, 'node_modules', 'mi/wasmnizer-ts');
    _sharedUtils.ColorConsole.info('### Waiting for generating app.wasm file.');
    const res = (0, _child_process.execSync)(`node ${ts2wasmpath}/build/cli/ts2wasm.js  ${entryFile} -o ${wasmDir}/app.wasm`, {
      encoding: 'utf-8'
    });
    _sharedUtils.ColorConsole.info(`### ts2wasm ### ${res.toString()}`);
  }
  if (compilerOption?.skip?.includes('ts2wasm')) {
    _sharedUtils.ColorConsole.info(`### skip compile ts to wasm due to --skip ${compilerOption?.skip.join(',')}`);
  } else {
    compile();
  }

  /** 2. 拷贝静态资源 */
  // TODO: ignore 数组待优化
  const files = _fastGlob.default.sync(`${_fastGlob.default.convertPathToPattern(projectPath)}/${output}/**/*`, {
    absolute: true,
    ignore: ['**/*.ts', '**/*.xts', '**/package.json', '**/package-lock.json', '**/*.wasm'],
    onlyFiles: true
  });
  for (const from of files) {
    const to = from.replace('/src', '').replace(`/${output}`, `/${wasmPackageName}`);
    _fs.default.mkdirSync(_path.default.dirname(to), {
      recursive: true
    });
    _fs.default.copyFileSync(from, to);
  }

  /** 3. 特殊处理 app/manifest.json */
  _fs.default.copyFileSync(_path.default.resolve(_fastGlob.default.convertPathToPattern(projectPath), output, 'app', 'manifest.json'), _path.default.resolve(_fastGlob.default.convertPathToPattern(projectPath), wasmPackageName, 'manifest.json'));
};
exports.ts2wasm = ts2wasm;
