// @aiot-toolkit/parser/lib/xts/validate/AttibuteValidate.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _sharedUtils = require("@aiot-toolkit/shared-utils");
var _CommonAttributeList = _interopRequireDefault(require("../config/CommonAttributeList"));
var _predefinedComponent = _interopRequireDefault(require("../config/predefinedComponent.json"));
var _AttributeUtil = _interopRequireDefault(require("../utils/AttributeUtil"));
var _TypescriptUtil = _interopRequireDefault(require("../utils/TypescriptUtil"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
/**
 * AttibuteValidate
 */
class AttibuteValidate {
  TRANSLATE_TYPE = {
    StringLiteral: 'string',
    NumbericLiteral: 'number',
    ObjectLiteralExpression: 'object'
  };
  validate(element, context, componentName) {
    const {
      sourceFile,
      onLog,
      offsetList
    } = context;
    // 属性节点值
    const elemName = element.name.getText();
    const elemValue = element.value.length > 0 ? element.value[0].getText() : '';
    const elemType = element.value[0];
    // 获取属性列表
    const attributeList = [..._CommonAttributeList.default];
    // 查看组件是否有特殊属性
    const componentGroups = _predefinedComponent.default.componentGroups;
    const compIndex = componentGroups.findIndex(comp => comp.name === componentName);
    const specialAttrList = compIndex !== -1 ? componentGroups[compIndex]?.attributeGroups : [];
    if (specialAttrList.length > 0) {
      attributeList.push(...specialAttrList);
    }
    // 检查属性是否正确
    const attrIndex = attributeList.findIndex(attr => attr.name === elemName);
    if (attrIndex !== -1) {
      const attr = attributeList[attrIndex];
      // 检查属性值\属性类型是否正确
      if (attr.type) {
        const errorStr = this.checkElemType(elemType, attr.type);
        errorStr && onLog({
          message: `Attribute ${elemName} ${errorStr}`,
          level: _sharedUtils.Loglevel.ERROR,
          position: _TypescriptUtil.default.getNodePostion(elemType, sourceFile, offsetList)
        });
      } else {
        // 判断elemValue,有值则报错，属性值不能为空
        elemValue && onLog({
          message: `Attribute ${elemName} value should be null`,
          level: _sharedUtils.Loglevel.ERROR,
          position: _TypescriptUtil.default.getNodePostion(element.name, sourceFile, offsetList)
        });
      }
    } else {
      // 组件上没有该属性
      onLog({
        message: `Component ${componentName} has no ${elemName} attibute`,
        level: _sharedUtils.Loglevel.ERROR,
        position: _TypescriptUtil.default.getNodePostion(element.name, sourceFile, offsetList)
      });
    }
    return [];
  }
  isISubTypeArray(instance) {
    return Array.isArray(instance);
  }
  /**
   * 检查属性类型是否正确
   * @param elemType
   * @param attrType
   */
  checkElemType(elemType, attrType) {
    if (this.isISubTypeArray(attrType)) {
      // 1. 静态数据时进行数据类型校验
      if (!_AttributeUtil.default.isDynamicValue(elemType)) {
        const typeList = attrType.map(item => item.subType);
        const type = this.TRANSLATE_TYPE[elemType.getKindName()];
        if (typeList.findIndex(item => item === type) === -1) {
          // 属性值应该为typeList中的一种
          return `value should be one of [${typeList.join(`,`)}]`;
        }
      }
    } else {
      // 属性的type类型应该为IType[]
      return 'type should be IType[]';
    }
    return '';
  }
}
var _default = exports.default = AttibuteValidate;
