// @aiot-toolkit/parser/lib/xts/XtsParser.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _sharedUtils = require("@aiot-toolkit/shared-utils");
var _tsMorph = require("ts-morph");
var _TypeUtil = _interopRequireDefault(require("./utils/TypeUtil"));
var _SourceFileValidate = _interopRequireDefault(require("./validate/SourceFileValidate"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
/**
 * xts生成 xts 的 ast 树
 * 工作内容为
 * 1. 替换component 为 class + 4个空格
 * 2. 使用typescript 解析成 ast 树
 */

class XtsParser {
  constructor(project, onLog) {
    this.project = project;
    this.onLog = _sharedUtils.WrapCallback.wrapOnLog(onLog);
  }
  parser(content, fileName) {
    const oldFile = this.project.getSourceFile('temp.xts');
    if (oldFile) {
      this.project.removeSourceFile(oldFile);
    }
    const sourceFile = this.project.createSourceFile('temp.xts', content);
    const offsetList = this.replaceKeyWord(sourceFile);
    // // 简单转换后，对内容进行错误检查
    // const reFile = this.project.createSourceFile('temp.ts', sourceFile.getFullText())
    // const fileDiags = reFile.getPreEmitDiagnostics().filter((item) => {
    //   return item.getCategory() === DiagnosticCategory.Error
    // })
    // if (fileDiags.length > 6) {
    //   throw new Error(`File have many error.\n${fileDiags.join('\n')}`)
    // }
    // if (this.project.getSourceFile('temp.ts')) {
    //   this.project.removeSourceFile(reFile)
    // }
    new _SourceFileValidate.default().validate({
      sourceFile,
      onLog: this.onLog,
      offsetList
    }, fileName);
    return {
      ast: sourceFile,
      offsetList
    };
  }
  replaceKeyWord(sourceFile) {
    const offsetList = [];
    let offsetTotal = 0;
    sourceFile.transform(traversal => {
      const node = traversal.visitChildren();
      if (_tsMorph.ts.isIdentifier(node)) {
        const text = node.getText();
        const matchedItem = _TypeUtil.default.getSpecialParentDecorator(text);
        if (matchedItem) {
          /**
           * 记录偏移位置, offset: 偏移量, pos：替换后新节点的结束位置
           * 例如：`app name` 替换为 `class name`
           * offset = app.length - class.length = -2，表示name在新代码的位置=6，则name在源位置=6+(-2) = 4
           * pos = 5，表示大于5的标识符，偏移量为-2
           */
          const newWord = matchedItem.replaceLabel;
          const currentOffset = text.length - newWord.length;
          offsetTotal += currentOffset;
          const nodePosition = sourceFile.getLineAndColumnAtPos(node.getStart());
          offsetList.push({
            offset: offsetTotal,
            line: nodePosition.line,
            column: nodePosition.column,
            pos: node.getEnd() - offsetTotal,
            currentOffset
          });
          return _tsMorph.ts.factory.createIdentifier(newWord);
        }
      }
      return node;
    });
    return offsetList;
  }
}
exports.default = XtsParser;
