// @aiot-toolkit/emulator/lib/instance/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "CommonInstance", {
  enumerable: true,
  get: function () {
    return _common.default;
  }
});
Object.defineProperty(exports, "GoldfishInstance", {
  enumerable: true,
  get: function () {
    return _dev.default;
  }
});
Object.defineProperty(exports, "MiwearInstance", {
  enumerable: true,
  get: function () {
    return _miwear.default;
  }
});
Object.defineProperty(exports, "PreInstance", {
  enumerable: true,
  get: function () {
    return _pre.default;
  }
});
Object.defineProperty(exports, "VelaPre5Instance", {
  enumerable: true,
  get: function () {
    return _pre2.VelaPre5Instance;
  }
});
exports.findInstance = findInstance;
var _common = _interopRequireDefault(require("./common"));
var _dev = _interopRequireDefault(require("./dev"));
var _miwear = _interopRequireDefault(require("./miwear"));
var _pre = _interopRequireDefault(require("./pre"));
var _Vvd = require("../typing/Vvd");
var _pre2 = require("./pre5");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
/**
 * 根据镜像决定使用哪个instance
 * Vela正式版(4.0) -> MiwearInstance
 * Vela正式版(不带 miwear 版本) -> PreInstance
 * Vela开发版（dev, 0.0.2) -> OldGoldfishInstance
 * Vela开发版（dev)，除0.0.2的其他版本 -> GoldfishInstance
 */
function findInstance(imageType, params) {
  const map = {
    [_Vvd.VelaImageType.PRE]: _pre.default,
    [_Vvd.VelaImageType.REL]: _miwear.default,
    [_Vvd.VelaImageType.DEV]: _dev.default,
    [_Vvd.VelaImageType.VELA_PRE_5]: _pre2.VelaPre5Instance
  };
  const Instance = map[imageType] || _dev.default;
  return new Instance(params);
}
