// @aiot-toolkit/emulator/lib/instance/miwear.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var adbMiwt = _interopRequireWildcard(require("@miwt/adb"));
var _path = _interopRequireDefault(require("path"));
var _utils = require("../utils");
var _common = _interopRequireDefault(require("./common"));
var _emulatorutil = require("../emulatorutil");
var _Vvd = require("../typing/Vvd");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/**
 * MiwearInstance
 * 针对 Vela正式版（4.0）的镜像
 */
class MiwearInstance extends _common.default {
  imageType = (() => _Vvd.VelaImageType.REL)();
  appDir = '/data/quickapp/app';

  /**
   * 使用 pm 安装快应用
   * @param targeRpk 快应用的rpk文件路径
   */
  async install(targeRpk) {
    adbMiwt.execAdbCmd(`adb -s ${this.sn} shell pm install ${targeRpk}`);
    return new Promise((resolve, reject) => {
      const func = msg => {
        if (_emulatorutil.EmulatorLog.inStallIsFinshe(msg)) {
          clearTimeout(timer);
          this.logger(`Install to ${this.vvdName} ${targeRpk} successfully`);
          resolve();
        }
      };
      let timer = setTimeout(() => {
        this.stdoutReadline.off('line', func);
        this.logger(`Install to ${this.vvdName} ${targeRpk} timeout`);
        reject('Install timeout');
      }, 2 * 60 * 1000);
      this.stdoutReadline.on('line', func);
    });
  }

  /** 使用 am start 启动快应用 */
  async startApp(packageName) {
    let debug = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : false;
    // 调试模式需要push一个文件至miwear中
    if (debug) {
      const debuggerCfgFile = _path.default.join(__dirname, '../static/debugger_ip.cfg');
      await adbMiwt.execAdbCmdAsync(`adb -s ${this.sn} push ${debuggerCfgFile} /data/debugger_ip.cfg`);
    } else {
      await adbMiwt.execAdbCmdAsync(`adb -s ${this.sn} shell rm /data/debugger_ip.cfg`);
    }
    const startCmd = `adb -s ${this.sn} shell am start ${packageName}`;
    this.logger(`Excuting: ${startCmd}`);
    await adbMiwt.execAdbCmdAsync(startCmd);
  }
  async closeApp(appName) {
    const stopCmd = `adb -s ${this.sn} shell am stop ${appName}`;
    await adbMiwt.execAdbCmdAsync(stopCmd);
    this.logger(`Stop ${this.vvdName} ${appName} successfully`);
    // 这里是为了等am stop命令清除资源等
  }
  async reboot() {
    await super.reboot();
    return new Promise((resolve, reject) => {
      const func = msg => {
        if (_emulatorutil.EmulatorLog.rpkIsStart(msg)) {
          clearTimeout(timer);
          resolve();
        }
      };
      let timer = setTimeout(() => {
        this.stdoutReadline.off('line', func);
        this.logger(`Reboot ${this.vvdName} timeout`);
        reject('reboot timeout');
      }, 2 * 60 * 1000);
      this.stdoutReadline.on('line', func);
    });
  }
  async reloadApp(appPackageName) {
    let _debug = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : false;
    try {
      // 2. 执行am stop和am start命令
      const stopCmd = `adb -s ${this.sn} shell am stop ${appPackageName}`;
      await adbMiwt.execAdbCmdAsync(stopCmd);
      this.logger(`Stop ${this.vvdName} ${appPackageName} successfully`);
      // 这里是为了等am stop命令清除资源等
      await (0, _utils.sleep)(500);
      const startCmd = `adb -s ${this.sn} shell am start ${appPackageName}`;
      await adbMiwt.execAdbCmdAsync(startCmd);
      this.logger(`Start ${this.vvdName} ${appPackageName} successfully`);
    } catch (e) {
      this.logger(`${e}`);
    }
  }
}
var _default = exports.default = MiwearInstance;
