// @aiot-toolkit/aiotpack/lib/loader/xts/XtsLoader.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _generator = require("@aiot-toolkit/generator");
var _parser = require("@aiot-toolkit/parser");
var _sharedUtils = require("@aiot-toolkit/shared-utils");
var _path = _interopRequireDefault(require("path"));
var _tsMorph = require("ts-morph");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
/**
 * XtsLoader
 */
class XtsLoader {
  parser(files) {
    const project = new _tsMorph.Project();
    const resultFiles = [];
    for (const file in files) {
      const {
        path,
        content
      } = files[file];
      const {
        name,
        ext
      } = _path.default.parse(path);
      const fullName = `${name}${ext}`;
      const newFileName = `${name}.ts`;
      //判断文本，空文本则结束该文件处理
      if (!content) {
        _sharedUtils.ColorConsole.warn(`The file '${fullName}' under the path '${path}' has no content `);
        return [];
      }
      const parserResult = new _parser.XtsParser(project, _sharedUtils.ColorConsole.logger).parser(content.toString(), fullName);
      const {
        ast,
        mapList
      } = new _parser.XtsToTypescript(project, _sharedUtils.ColorConsole.logger, this.context, path).translate(parserResult.ast, parserResult.offsetList);
      const {
        code,
        sourcemap
      } = new _generator.TypescriptGenerator().generate({
        sourceFilePath: fullName,
        targetFilePath: newFileName,
        ast,
        mapList
      });
      resultFiles.push({
        path: _sharedUtils.FileUtil.updateFileName(path, newFileName),
        content: code
      }, {
        path: _sharedUtils.FileUtil.updateFileName(path, `${name}.map.ts`),
        content: sourcemap
      });
    }
    // 返回内容
    return resultFiles;
  }
}
var _default = exports.default = XtsLoader;
