// @aiot-toolkit/emulator/lib/typing/Vvd.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.VelaImageType = exports.VELAHOME = exports.SDKParts = exports.IVvdArchType = void 0;
let IVvdArchType = exports.IVvdArchType = /*#__PURE__*/function (IVvdArchType) {
  IVvdArchType["arm"] = "arm";
  IVvdArchType["arm64"] = "arm64";
  return IVvdArchType;
}({});
let VelaImageType = exports.VelaImageType = /*#__PURE__*/function (VelaImageType) {
  VelaImageType["VELA_PRE_5"] = "vela-pre-5.0";
  VelaImageType["REL"] = "vela-release-4.0";
  VelaImageType["PRE"] = "vela-pre-4.0";
  VelaImageType["DEV"] = "vela-dev-0.0.4";
  return VelaImageType;
}({});
let SDKParts = exports.SDKParts = /*#__PURE__*/function (SDKParts) {
  SDKParts["EMULATOR"] = "emulator";
  SDKParts["QA"] = "qa";
  SDKParts["SKINS"] = "skins";
  SDKParts["SYSTEM_IMAGES"] = "system-images";
  SDKParts["MODEM_SIMULATOR"] = "modem_simulator";
  return SDKParts;
}({});
let VELAHOME = exports.VELAHOME = /*#__PURE__*/function (VELAHOME) {
  VELAHOME["SDK"] = ".export";
  VELAHOME["VVD"] = ".vela/vvd";
  return VELAHOME;
}({});
