// @aiot-toolkit/parser/lib/xts/enum/AttributeType.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
/**
 * AttributeType
 */
var AttributeType = /*#__PURE__*/function (AttributeType) {
  AttributeType[AttributeType["EVENT"] = 1] = "EVENT";
  AttributeType[AttributeType["ATTRIBUTE"] = 2] = "ATTRIBUTE";
  AttributeType[AttributeType["STYLE"] = 3] = "STYLE";
  AttributeType[AttributeType["FOCUSEDSTYLE"] = 4] = "FOCUSEDSTYLE";
  AttributeType[AttributeType["PRESSEDSTYLE"] = 5] = "PRESSEDSTYLE";
  AttributeType[AttributeType["DISABLEDSTYLE"] = 6] = "DISABLEDSTYLE";
  return AttributeType;
}(AttributeType || {});
(function (_AttributeType) {
  const ImageComponet = _AttributeType.ImageComponet = ['Image', 'ImageAnimator', 'backgroundImage'];
  const ImageAttrs = _AttributeType.ImageAttrs = ['background-image', 'backgroundImage'];
  let StyleNodeList = /*#__PURE__*/function (StyleNodeList) {
    StyleNodeList["STYLE"] = "style";
    StyleNodeList["FOCUSEDSTYLE"] = "focusedStyle";
    StyleNodeList["PRESSEDSTYLE"] = "pressedStyle";
    StyleNodeList["DISABLEDSTYLE"] = "disabledStyle";
    return StyleNodeList;
  }({});
  _AttributeType.StyleNodeList = StyleNodeList;
  const ALL_LIST = _AttributeType.ALL_LIST = [AttributeType.ATTRIBUTE, AttributeType.EVENT, AttributeType.STYLE, AttributeType.FOCUSEDSTYLE, AttributeType.PRESSEDSTYLE, AttributeType.DISABLEDSTYLE];
  function toString(value) {
    const dic = {
      [AttributeType.EVENT]: 'event',
      [AttributeType.ATTRIBUTE]: 'attr',
      [AttributeType.STYLE]: 'style',
      [AttributeType.FOCUSEDSTYLE]: 'focusedStyle',
      [AttributeType.PRESSEDSTYLE]: 'pressedStyle',
      [AttributeType.DISABLEDSTYLE]: 'disabledStyle'
    };
    return dic[value] || '';
  }
  _AttributeType.toString = toString;
  function isStyle(attr) {
    return {
      [StyleNodeList.STYLE]: StyleNodeList.STYLE,
      [StyleNodeList.FOCUSEDSTYLE]: StyleNodeList.FOCUSEDSTYLE,
      [StyleNodeList.PRESSEDSTYLE]: StyleNodeList.PRESSEDSTYLE,
      [StyleNodeList.DISABLEDSTYLE]: StyleNodeList.DISABLEDSTYLE
    }[attr];
  }
  _AttributeType.isStyle = isStyle;
})(AttributeType || (AttributeType = {}));
var _default = exports.default = AttributeType;
