// @aiot-toolkit/parser/lib/xts/enum/Framework.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.tsFrameWorkExports = exports.tsFrameWork = exports.NodeStyles = exports.NodeAttrs = exports.NativeFeatureKey = exports.HomeState = void 0;
// Don't edit, generate by scripts/buildXtsApi.ts
const HomeState = exports.HomeState = {
  interface: 'HomeState_'
};
/**
  * feature keys
  */
class NativeFeatureKey {
  static SYSTEM_CONFIGURATION = 'mios.configuration';
  static SYSTEM_STORAGE = 'mios.storage';
  static SYSTEM_DEVICE = 'mios.device';
  static SYSTEM_FILE = 'mios.file';
  static SYSTEM_CIPHER = 'mios.cipher';
  static SYSTEM_BRIGHTNESS = 'mios.brightness';
  static SYSTEM_FETCH = 'mios.fetch';
  static SYSTEM_REQUEST = 'mios.request';
  static SYSTEM_INTERCONNECT = 'mios.interconnect';
  static SYSTEM_AUDIO = 'mios.audio';
  static SYSTEM_NETWORK = 'mios.network';
  static SYSTEM_BATTERY = 'mios.battery';
  static SYSTEM_SENSOR = 'mios.sensor';
  static SYSTEM_VIBRATOR = 'mios.vibrator';
  static SYSTEM_PROMPT = 'mios.prompt';
  static SYSTEM_RECORD = 'mios.record';
  static SERVICE_HEALTH = 'mios.health';
  static SERVICE_WECHAT = 'mios.wechat';
  static SYSTEM_CRYPTO = 'mios.crypto';
  static SYSTEM_VOLUME = 'mios.volume';
  static SYSTEM_EXCHANGE = 'mios.exchange';
  static SYSTEM_GEOLOCATION = 'mios.geolocation';
  static SYSTEM_ROUTER = 'mios.router';
  static SYSTEM_APP = 'mios.app';
}
exports.NativeFeatureKey = NativeFeatureKey;
const NodeAttrs = exports.NodeAttrs = ['alt', 'autoplay', 'bounces', 'checked', 'dataset', 'datasets', 'disabled', 'direction', 'duration', 'enableswipe', 'fillmode', 'fixedsize', 'href', 'id', 'images', 'index', 'indicator', 'interval', 'iteration', 'loop', 'max', 'min', 'name', 'nextmargin', 'options', 'percent', 'previousmargin', 'reverse', 'selected', 'scrollamount', 'scrollX', 'scrollY', 'scrollTop', 'scrollBottom', 'scrollLeft', 'scrollRight', 'show', 'spinnerStyle', 'src', 'start', 'step', 'type', 'value', 'vertical'];
const NodeStyles = exports.NodeStyles = ['alignItems', 'backgroundImage', 'blockColor', 'border', 'borderStyle', 'borderRadius', 'borderWidth', 'bottom', 'display', 'flex', 'flexBasis', 'flexDirection', 'flexGrow', 'flexShrink', 'height', 'justifyContent', 'left', 'marginBottom', 'marginLeft', 'marginRight', 'marginTop', 'minWidth', 'minHeight', 'maxWidth', 'maxHeight', 'opacity', 'paddingBottom', 'paddingLeft', 'paddingRight', 'paddingTop', 'position', 'right', 'top', 'visibility', 'width', 'animationDelay', 'animationDirection', 'animationDuration', 'animationFillMode', 'animationIterationCount', 'animationName', 'animationTimingFunction', 'autoplay', 'borderBottomColor', 'borderBottomWidth', 'borderLeftColor', 'borderLeftWidth', 'borderRightColor', 'borderRightWidth', 'borderTopColor', 'borderTopWidth', 'canvas', 'caretColor', 'center', 'centerX', 'centerY', 'column', 'columns', 'columnReverse', 'columnSpan', 'change', 'checked', 'click', 'clip', 'datasets', 'date', 'direction', 'drag', 'duration', 'easeIn', 'easeInOut', 'easeOut', 'ellipsis', 'expand', 'flexEnd', 'flexStart', 'flexWrap', 'forwards', 'index', 'indicatorColor', 'indicatorSelectedColor', 'indicatorSize', 'indicatorTop', 'indicatorLeft', 'indicatorRight', 'indicatorBottom', 'itemselected', 'iteration', 'key', 'labels', 'larger', 'layerColor', 'layoutType', 'letterSpacing', 'lineHeight', 'lines', 'longpress', 'loop', 'marquee', 'maskColor', 'max', 'min', 'muted', 'name', 'none', 'objectFit', 'options', 'overflow', 'pageAnimationName', 'pageTransformOrigin', 'pause', 'percent', 'placeholderColor', 'progressColor', 'radius', 'range', 'ref', 'reverse', 'rotate', 'row', 'rowReverse', 'scaleX', 'scaleY', 'scrollamount', 'scrollend', 'scrollstart', 'scrollSnapType', 'scrollSnapAlign', 'seeked', 'seeking', 'selected', 'selectedColor', 'selectedFontSize', 'show', 'spaceAround', 'spaceBetween', 'start', 'startAngle', 'standard', 'stop', 'strokeWidth', 'styleList', 'textAlign', 'textDecoration', 'textIndent', 'textOverflow', 'thumbColor', 'totalAngle', 'trackColor', 'transform', 'transformOrigin', 'translateX', 'translateY', 'transitionDelay', 'transitionDuration', 'transitionProperty', 'transitionTimingFunction', 'value', 'vertical', 'wrap'];
let tsFrameWorkExports = exports.tsFrameWorkExports = /*#__PURE__*/function (tsFrameWorkExports) {
  tsFrameWorkExports["App"] = "App";
  tsFrameWorkExports["AppContainer"] = "AppContainer";
  tsFrameWorkExports["AppDefinition"] = "AppDefinition";
  tsFrameWorkExports["AspectDefinition"] = "AspectDefinition";
  tsFrameWorkExports["Aspect"] = "Aspect";
  tsFrameWorkExports["ViewAspect"] = "ViewAspect";
  tsFrameWorkExports["Component"] = "Component";
  tsFrameWorkExports["Page"] = "Page";
  tsFrameWorkExports["PageDefinition"] = "PageDefinition";
  tsFrameWorkExports["VTag"] = "VTag";
  tsFrameWorkExports["stats"] = "stats";
  tsFrameWorkExports["context"] = "context";
  tsFrameWorkExports["AppLifecycle"] = "AppLifecycle";
  tsFrameWorkExports["AspectLifecycle"] = "AspectLifecycle";
  tsFrameWorkExports["Route"] = "Route";
  tsFrameWorkExports["RouteInfo"] = "RouteInfo";
  tsFrameWorkExports["AppManifest"] = "AppManifest";
  tsFrameWorkExports["AspectManifest"] = "AspectManifest";
  tsFrameWorkExports["IComponentState"] = "IComponentState";
  tsFrameWorkExports["PlainObject"] = "PlainObject";
  tsFrameWorkExports["getAppManifest"] = "getAppManifest";
  tsFrameWorkExports["getAspectManifest"] = "getAspectManifest";
  tsFrameWorkExports["ImageFrame"] = "ImageFrame";
  tsFrameWorkExports["ChartAxis"] = "ChartAxis";
  tsFrameWorkExports["ChartLineStyle"] = "ChartLineStyle";
  tsFrameWorkExports["ChartLoop"] = "ChartLoop";
  tsFrameWorkExports["ChartSeries"] = "ChartSeries";
  tsFrameWorkExports["ChartOptions"] = "ChartOptions";
  tsFrameWorkExports["ChartDataset"] = "ChartDataset";
  tsFrameWorkExports["NodeAttrs"] = "NodeAttrs";
  tsFrameWorkExports["on"] = "on";
  tsFrameWorkExports["getInfoFromRoute"] = "getInfoFromRoute";
  tsFrameWorkExports["EventPhase"] = "EventPhase";
  tsFrameWorkExports["Event"] = "Event";
  tsFrameWorkExports["BubbleEvent"] = "BubbleEvent";
  tsFrameWorkExports["MouseEvent"] = "MouseEvent";
  tsFrameWorkExports["TouchListItem"] = "TouchListItem";
  tsFrameWorkExports["TouchEvent"] = "TouchEvent";
  tsFrameWorkExports["SwipeEvent"] = "SwipeEvent";
  tsFrameWorkExports["ScrollEvent"] = "ScrollEvent";
  tsFrameWorkExports["ListScrollEvent"] = "ListScrollEvent";
  tsFrameWorkExports["SwiperEvent"] = "SwiperEvent";
  tsFrameWorkExports["CompleteEvent"] = "CompleteEvent";
  tsFrameWorkExports["InputChangeEvent"] = "InputChangeEvent";
  tsFrameWorkExports["RadioChangeEvent"] = "RadioChangeEvent";
  tsFrameWorkExports["CheckboxChangeEvent"] = "CheckboxChangeEvent";
  tsFrameWorkExports["PickerChangeEvent"] = "PickerChangeEvent";
  tsFrameWorkExports["TimePickerChangeEvent"] = "TimePickerChangeEvent";
  tsFrameWorkExports["SwitchChangeEvent"] = "SwitchChangeEvent";
  tsFrameWorkExports["SliderChangeEvent"] = "SliderChangeEvent";
  tsFrameWorkExports["CustomEvent"] = "CustomEvent";
  tsFrameWorkExports["requireFeature"] = "requireFeature";
  tsFrameWorkExports["registerFeatures"] = "registerFeatures";
  tsFrameWorkExports["unRegisterFeatures"] = "unRegisterFeatures";
  tsFrameWorkExports["MiosRouter"] = "MiosRouter";
  tsFrameWorkExports["MiosStorage"] = "MiosStorage";
  tsFrameWorkExports["MiosFetch"] = "MiosFetch";
  tsFrameWorkExports["SuccessRes"] = "SuccessRes";
  return tsFrameWorkExports;
}({});
const tsFrameWork = exports.tsFrameWork = {
  name: 'mi/ts-framework',
  exports: tsFrameWorkExports
};
