// aiot-toolkit/lib/builder/VelaUxBuilder.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.velaUxBuilderParams = exports.default = void 0;
var _sharedUtils = require("@aiot-toolkit/shared-utils");
var _UxBuilderBase = _interopRequireDefault(require("./UxBuilderBase"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
const velaUxBuilderParams = exports.velaUxBuilderParams = [{
  type: 'string',
  name: 'devtool',
  description: 'source map config'
}, {
  name: 'disable-subpackages',
  description: 'disable subpackages'
}, {
  type: 'confirm',
  name: 'enable-jsc',
  description: 'enable jsc bundle',
  defaultValue: false
}, {
  type: 'confirm',
  name: 'enable-protobuf',
  description: 'enable protobuf',
  defaultValue: false
}, {
  type: 'confirm',
  name: 'enable-e2e',
  description: 'inject test-suite for current project',
  defaultValue: false
}, {
  type: 'confirm',
  name: 'enable-stats',
  description: 'analyse time and size of webpack output files',
  defaultValue: false
}, {
  type: 'confirm',
  name: 'optimize-css-attr',
  description: 'optimize css attr',
  defaultValue: false
}, {
  type: 'confirm',
  name: 'complete-feature',
  description: 'auto complete the features are used in the code into the features field of the manifest.json file',
  defaultValue: false
}, {
  type: 'confirm',
  name: 'enable-custom-component',
  description: 'deprecated',
  deprecated: 'please remove it now',
  deprecatedVersion: '2.1.0'
}, {
  type: 'confirm',
  name: 'stats',
  description: 'deprecated',
  deprecated: 'please use `enable-stats` instead',
  deprecatedVersion: '2.1.0'
}];

/**
 * VelaUxBuilder
 */
class VelaUxBuilder extends _UxBuilderBase.default {
  params = (() => velaUxBuilderParams)();
  match(projectPath) {
    return _sharedUtils.ProjectType.getProjectType(projectPath) === _sharedUtils.ProjectType.VELA_UX;
  }
}
var _default = exports.default = VelaUxBuilder;
