// aiot-toolkit/lib/builder/XtsBuilder.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _aiotpack = require("@aiot-toolkit/aiotpack");
var _sharedUtils = require("@aiot-toolkit/shared-utils");
var _fileLane = require("file-lane");
var _fs = _interopRequireDefault(require("fs"));
var _path = _interopRequireDefault(require("path"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
/**
 * XtsBuilder
 */
class XtsBuilder {
  static PROJECT_TYPE = 'xts quick app';
  match(projectPath) {
    // app/app.xts 存在视为xts项目
    return _fs.default.existsSync(_path.default.resolve(projectPath, 'app/app.xts'));
  }
  async build(projectPath, options) {
    if (this.fileLane) {
      // 已经在运行中，请先停止
      throw new Error('already running, please stop first: builder.stop()');
    }
    const watch = options.watch || false;
    _sharedUtils.ColorConsole.success('Start build...\n', {
      word: 'ProjectPath: ',
      style: _sharedUtils.ColorConsole.getStyle(_sharedUtils.Loglevel.SUCCESS)
    }, projectPath, '\n', {
      word: 'buildOptions: ',
      style: _sharedUtils.ColorConsole.getStyle(_sharedUtils.Loglevel.SUCCESS)
    }, JSON.stringify(options));
    const compilerOptions = {
      skip: _sharedUtils.StringUtil.string2arrayByComma(options.skip)
    };
    const config = new _aiotpack.XtsConfig(projectPath);
    if (compilerOptions.skip?.includes('xts2ts')) {
      _sharedUtils.ColorConsole.info("### skip compile xts to ts due to --skip ${compilerOptions?.skip.join(',')}");
      const context = _fileLane.FileLaneUtil.createContext(config.output, projectPath);
      const preWorks = config.beforeCompile || [];
      for (let item of preWorks) {
        try {
          await item({
            context,
            config,
            compilerOption: compilerOptions
          });
        } catch (error) {
          // 报错 prework的item error
        }
      }
      const follWorks = config.afterCompile || [];
      for (let item of follWorks) {
        try {
          await item.worker({
            context,
            config,
            compilerOption: compilerOptions
          });
        } catch (error) {
          // 报错 prework的item error
        }
      }
    } else {
      this.fileLane = new _fileLane.FileLane(config, projectPath, compilerOptions);
      this.fileLane.start({
        watch
      });
    }
  }
  async dispose() {
    await this.fileLane?.dispose();
    this.fileLane = undefined;
  }
  params = (() => [{
    name: 'skip',
    type: 'string',
    description: `Can configure skip steps, comma separated, optional values: ${_aiotpack.skipList.join(',')}`,
    validate(value) {
      // TODO: validate 不起作用
      const res = _sharedUtils.StringUtil.string2arrayByComma(value);
      const unValid = res.find(r => {
        return !_aiotpack.skipList.includes(r);
      });
      if (unValid) {
        return `${unValid} is unvalidate,  validate value are ${_aiotpack.skipList.join(',')}, Multiple values separated by commas`;
      }
      return true;
    }
  }])();
}
var _default = exports.default = XtsBuilder;
