// @aiot-toolkit/parser/lib/xts/utils/AttributeUtil.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _sharedUtils = require("@aiot-toolkit/shared-utils");
var _path = _interopRequireDefault(require("path"));
var _tsMorph = require("ts-morph");
var _AttributeType = _interopRequireDefault(require("../enum/AttributeType"));
var _Framework = require("../enum/Framework");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
/**
 * AttributeUtil
 */
class AttributeUtil {
  /**
   * 通过属性名称，判断属性类型
   * @param name 属性名称
   */
  static getAttributeType(name) {
    // 临时规则，需要改成从配置文件中读取
    // 如果是 on 开头的，是 event
    // 如果是 width 是style
    // 其它是 attr
    if (name.startsWith('on')) {
      return _AttributeType.default.EVENT;
    } else if (_Framework.NodeStyles.includes(name)) {
      return _AttributeType.default.STYLE;
    }
    return _AttributeType.default.ATTRIBUTE;
  }

  /**
   * 是否是静态属性
   * @param value
   * @returns
   */
  static isDynamicValue(value) {
    if ([_tsMorph.SyntaxKind.StringLiteral, _tsMorph.SyntaxKind.NumericLiteral].includes(value.getKind())) return false;else if (value.getKind() === _tsMorph.SyntaxKind.ObjectLiteralExpression) {
      const props = value.getProperties();
      let hasDynamic = false;
      for (const p of props) {
        hasDynamic = AttributeUtil.isDynamicValue(p.getInitializer());
        if (hasDynamic) break;
      }
      return hasDynamic;
    }
    return true;
  }

  /**
   * 根据 表达式式节点的内容，向 object对象 添加属性
   *
   * @description 根据  width(100) 向  {} 添加 style: {width: 100 } 属性
   * 1. 通过 width 判断属性为应该是 style
   * 2. 取得 width 的值100
   * 3. 修改 {} 为 { style: {width: 100 }}
   *
   * @param expressNode
   * @param target
   * @returns
   */
  static addAttribute(expressNode, map, index, fileDir, root, pushXtsImport) {
    const styleType = _AttributeType.default.isStyle(expressNode.name.getText());
    if (styleType) {
      return AttributeUtil.addStyle(expressNode, styleType, map, index, fileDir, root);
    }
    const type = this.getAttributeType(expressNode.name.getText());
    if (type === _AttributeType.default.EVENT) {
      return AttributeUtil.addEvent(expressNode, map, index, fileDir, root, pushXtsImport);
    }
    const attributeName = _AttributeType.default.toString(type);
    const key = AttributeUtil.createVarName(attributeName, index);
    const variableStatement = map.get(key);
    const obj = variableStatement.getDeclarations()[0];
    const value = expressNode.value[0];
    let initStr = value ? value.getText() : '';

    // 如果是 图片相关的属性，需要进行转换
    if (_AttributeType.default.ImageAttrs.includes(expressNode.name.getText())) {
      initStr = AttributeUtil.transformSrc(initStr, fileDir, root);
    }
    const attributeItem = AttributeUtil.assignProp(obj.getInitializer(), {
      name: expressNode.name.getText(),
      initializer: initStr
    });
    return {
      type: attributeName,
      attributeItem,
      isDynamic: value ? this.isDynamicValue(value) : false
    };
  }
  static addEvent(expressNode, map, index, fileDir, root, pushXtsImport) {
    const attributeName = _AttributeType.default.toString(_AttributeType.default.EVENT);
    const key = AttributeUtil.createVarName(attributeName, index);
    const variableStatement = map.get(key);
    const obj = variableStatement.getDeclarations()[0];
    const isCustomEvnet = expressNode.name.getText() === 'onEvent';
    const funcNode = isCustomEvnet ? expressNode.value[1] : expressNode.value[0];
    const eventName = isCustomEvnet ? expressNode.value[0].getText() : expressNode.name.getText();
    const eventType = funcNode.getParameters && funcNode.getParameters()?.at(0)?.getTypeNode()?.getText() || 'PlainObject';
    let callParams = '';
    if (!isCustomEvnet) {
      callParams = `e__ as ${eventType}`;
      pushXtsImport(eventType);
    } else if (isCustomEvnet && funcNode.getParameters() && funcNode.getParameters().length > 0) {
      callParams = '(e__ as CustomEvent).data';
      pushXtsImport(_Framework.tsFrameWork.exports.CustomEvent);
    }
    let initStr = funcNode ? `(e__: Event) => {
        (${funcNode.getText()})(${callParams})
      }` : '';
    pushXtsImport(_Framework.tsFrameWork.exports.Event);

    // 如果是 图片相关的属性，需要进行转换
    if (_AttributeType.default.ImageAttrs.includes(expressNode.name.getText())) {
      initStr = AttributeUtil.transformSrc(initStr, fileDir, root);
    }
    const attributeItem = AttributeUtil.assignProp(obj.getInitializer(), {
      name: eventName,
      initializer: initStr
    });
    return {
      type: attributeName,
      attributeItem,
      isDynamic: funcNode ? this.isDynamicValue(funcNode) : false
    };
  }

  /**
   * 根据表达式式节点的内容，向 object 对象添加属性
   *
   * @description 根据  style({backgrourd:'red'}) 向  {} 添加 style: { backgrourd:'red' } 属性
   *
   * @param expressNode
   * @param target
   * @returns
   */
  static addStyle(expressNode, styleType, map, index, fileDir, root) {
    const key = AttributeUtil.createVarName(styleType, index);
    const variableStatement = map.get(key);
    const obj = variableStatement.getDeclarations()[0];
    const value = expressNode.value[0]
    // TODO: value 有可能不是对象的情况
    ;
    value.getProperties().forEach(prop => {
      const oldStyle = obj.getInitializer();
      let initializer = prop.getInitializer().getText();
      // 如果是 图片相关的属性，需要进行转换
      if (_AttributeType.default.ImageAttrs.includes(_sharedUtils.StringUtil.removeQuotes(prop.getName()))) {
        initializer = AttributeUtil.transformSrc(initializer, fileDir, root);
      }
      AttributeUtil.assignProp(oldStyle, {
        name: prop.getName(),
        initializer
      });
    });
    return {
      type: styleType,
      attributeItem: obj,
      isDynamic: value ? this.isDynamicValue(value) : false
    };
  }

  /** 给一个对象节点添加属性，若已存在则删除原属性后添加 */
  static assignProp(tar, structure) {
    const allProps = tar.getProperties();
    for (const p of allProps) {
      if (p.getName() === structure.name) {
        p.remove();
        break;
      }
    }
    return tar.addPropertyAssignment(structure);
  }
  static createVarName(type, index) {
    return `${type}_${index}`;
  }
  static reverseVarName(name) {
    return name.replace(/_\d/, '');
  }

  /**
   * 将图片的 src 相对路径改为从根路径开始的绝对路径
   * @param src 用户设置的 src
   * @param fileDir 该文件的目录
   * @param root 项目根目录
   */
  static transformSrc(src, fileDir, root) {
    // 网络地址不用处理
    if (src.startsWith('http')) return src;

    // 绝对路径
    const p1 = _path.default.resolve(fileDir, _sharedUtils.StringUtil.removeQuotes(src));
    return `'${p1.replace(root, '').replace('/src', '')}'`;
  }
  static getFrameworkImportFrom(srouceFilePath, root) {
    const frameworkPath = require.resolve(_Framework.tsFrameWork.name, {
      paths: [root]
    });
    const res = _path.default.relative(srouceFilePath, frameworkPath).replace(/\\/g, '/');
    return res.replace('.ts', '');
  }
}
var _default = exports.default = AttributeUtil;
