// @aiot-toolkit/parser/lib/xts/utils/TypescriptUtil.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _sharedUtils = require("@aiot-toolkit/shared-utils");
var _tsMorph = require("ts-morph");
var _ParserUtil = _interopRequireDefault(require("../../utils/ParserUtil"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
/**
 * TypescriptUtil
 */
class TypescriptUtil {
  /**
   * 拆分表达式为多个节点列表
   *
   * 一个表达式可能包含多项值，且是反向的，需要分离并把顺序调正
   *
   * 例如 a().width(100).height(200) 1个表达式包含3项，且第1项是 height，最后一项是 a；需要分离为 [a, width, height]
   * @param node
   * @returns
   */
  static splitExpression(expression) {
    const result = [];
    let currentItem = expression.getExpression();
    while (currentItem) {
      if (currentItem.getKind() === _tsMorph.SyntaxKind.CallExpression) {
        const express = currentItem.getExpression();
        // 属性节点 row().width()中的 width()
        if (express instanceof _tsMorph.PropertyAccessExpression) {
          result.push({
            name: express.getNameNode(),
            value: currentItem.getArguments()
          });
        }
        // 起始节点  row().width()中的 row()
        else if (_tsMorph.ts.isIdentifier(express.compilerNode)) {
          result.push({
            name: express,
            value: currentItem.getArguments()
          });
        }
      }
      if (currentItem.getExpression) {
        currentItem = currentItem.getExpression();
      } else {
        currentItem = undefined;
      }
    }
    return result.reverse();
  }
  static throwError(message, node, sourceFile, offsetList) {
    let str = message;
    if (node) {
      const position = sourceFile ? this.getNodePostion(node, sourceFile, offsetList) : undefined;
      str = JSON.stringify({
        message,
        code: node.getText(),
        position
      }, undefined, 2);
      if (position) {
        throw new _sharedUtils.PositionError(position, str);
      }
    }
    throw new Error(str);
  }

  /**
   * 获取目标代码的节点在源码中的位置
   * @param node 节点
   * @param sourceFile 目标文件对象
   * @param offsetList 源码相对目标代码的偏移量列表
   * @returns
   */
  static getNodePostion(node, sourceFile) {
    let offsetList = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : [];
    const start = node.getStart();
    const end = node.getEnd();
    const offset = offsetList ? _ParserUtil.default.getOffsetPos(start, offsetList) : 0;
    const {
      line: startLine,
      column: startColumn
    } = sourceFile.getLineAndColumnAtPos(start);
    const {
      line: endLine,
      column: endColumn
    } = sourceFile.getLineAndColumnAtPos(end);
    const columnOffset = _ParserUtil.default.getOffsetColumn({
      line: startLine,
      pos: start
    }, offsetList);
    return {
      pos: start + offset,
      end: end + offset,
      startLine,
      startColumn: startColumn + columnOffset,
      endLine,
      endColumn: endColumn + columnOffset
    };
  }

  /**
   * 通过 getChildren 方法获取 block 的子元素
   *
   * 例如，下面的代码，需要获取 Row 的子元素列表
   * ```
   * Row(){
   *   Button().x(10)
   * }
   * ```
   * @param block
   * @returns
   */
  static getBlockChildren(block) {
    if (!block) {
      return [];
    }
    const blockChildren = block.getChildren();
    // 获取SyntaxKind.SyntaxList的节点，并获取它的children
    const syntaxListNode = blockChildren.find(item => item.getKind() === _tsMorph.SyntaxKind.SyntaxList);
    if (syntaxListNode) {
      return syntaxListNode.getChildren();
    }
    return [];
  }
}
var _default = exports.default = TypescriptUtil;
