// @aiot-toolkit/parser/lib/xts/utils/TypeUtil.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
/**
 * TypeUtil
 */
class TypeUtil {
  static SIMPLE_DATA_TYPE = ['string', 'number', 'boolean', 'Date'];
  static PRIMAITIVE_TYPES = ['number', 'string', 'boolean', 'null', 'undefined', 'symbol'];
  static COMPLEX_DTAT_TYPE = ['Array', 'Map', 'Set'];
  static SPECIAL_DECORATORS = ['@State', '@Prop', '@PropRequired'];
  static SPECIAL_PARENT_DECORATORS = [{
    label: 'app',
    replaceLabel: `@parentClass(App) class`
  }, {
    label: 'vaspect',
    replaceLabel: `@parentClass(ViewAspect) class`
  }, {
    label: 'component',
    replaceLabel: `@parentClass(Component) class`
  }];

  /**
   * 判断类型是否是 ts 原始类型
   * @param type
   * @returns
   */
  static isPrimitiveType(type) {
    if (!type) return false;
    return this.PRIMAITIVE_TYPES.includes(type);
  }

  /**
   * 判断数据类型是否为简单类型
   * @param dataType
   * @returns
   */
  static isSimpleType(dataType) {
    if (!dataType) return;
    return this.SIMPLE_DATA_TYPE.includes(dataType);
  }
  /**
   * 判断数据类型是否为自定义类型
   * @param dataType
   * @returns
   */
  static isCustomType(dataType) {
    const name = dataType?.getTypeName().getText();
    return !this.COMPLEX_DTAT_TYPE.includes(name);
  }
  /**
   * 判断类节点是否为一般类
   * @param sourceNode
   * @returns
   */
  static isNormalClass(sourceNode) {
    const decorators = sourceNode.getDecorators();
    return !decorators.find(decorator => {
      const decoratorContent = decorator.getText();
      return this.SPECIAL_PARENT_DECORATORS.some(item => item.replaceLabel === `${decoratorContent} class`);
    });
  }
  /**
   * 判断节点identifier是否为特殊类关键词，并返回匹配项
   * @param sourceText
   * @returns
   */
  static getSpecialParentDecorator(sourceText) {
    const matched = this.SPECIAL_PARENT_DECORATORS.find(item => item.label === sourceText);
    return matched;
  }
  /**
   * 获取数据节点的特定修饰符
   * @param node
   * @returns
   */
  static getSpecialDecorator(node) {
    const decorators = node.getDecorators();
    const specialDecoratorNode = decorators.find(dec => {
      return this.SPECIAL_DECORATORS.includes(dec.getText());
    });
    return specialDecoratorNode?.getText();
  }
  /**
   * 获取数据节点的特定修饰符是否为'@Prop', '@PropRequired'
   * @param node
   * @returns
   */
  static isDecoratorProps(node) {
    const spDecoratorStr = this.getSpecialDecorator(node);
    return spDecoratorStr && spDecoratorStr !== this.SPECIAL_DECORATORS[0];
  }
}
var _default = exports.default = TypeUtil;
