// @aiot-toolkit/aiotpack/lib/loader/ux/JsLoader.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _parser = require("@aiot-toolkit/parser");
var _sharedUtils = require("@aiot-toolkit/shared-utils");
/**
 * JsLoader
 */
class JsLoader {
  parser(files) {
    const onLog = () => {};
    return files.map(item => {
      if (!item.content) {
        return {
          path: item.path,
          content: item.content
        };
      }
      const options = {
        filePath: item.path,
        projectPath: this.context.projectPath,
        content: item.content.toString(),
        projectType: _sharedUtils.ProjectType.getProjectType(this.context.projectPath),
        onLog
      };
      return {
        path: item.path,
        content: new _parser.ScriptToTypescript(options, this.compilerOption, this.context).translate({
          content: new _parser.ScriptParser(options).parser(item.content.toString()).ast.content
        }, []).targetTree.getFullText()
      };
    });
  }
}
var _default = exports.default = JsLoader;
