// @aiot-toolkit/parser/lib/ux/utils/StyleMapUtil.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _path = _interopRequireDefault(require("path"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
class StyleMapUtil {
  constructor(sourceMapConsumer, lineOffset, options) {
    this.sourceMapConsumer = sourceMapConsumer;
    this.lineOffset = lineOffset;
    this.options = options;
  }
  /**
   * 1. 根据生成结果位置找转换前的行列值
   * 2. 根据样式代码总体偏移量，计算出当前节点行在ux中的行
   *  2.1 查看map信息，原始行总是从第三行开始计算，-2是去掉生成map信息时多余的偏移量
   *  2.2 再加上样式代码总体偏移量，就可以计算出实际所在行数
   * 3. 返回节点原始位置信息
   *  3.1 原始位置的startLine即为上一步计算了偏移量的行信息
   *  3.2 原始位置的startColumn为查找到的节点列信息（目前看结果不存在列偏移）
   *  3.3 原始位置的endLine为startLine+生成结果中所占据行数的值
   *  3.4 原始位置的endColumn根据map结果，发现存在1个位置的差异，所以-1
   *  3.5 source存储来源的文件
   * @param location 生成结果中节点的位置信息
   * @param consumer 经过解析器获取的map信息生成的consumer对象
   * @param offset 当前样式代码在ux文件中的偏移量
   * @returns
   */
  transfromLocToPosition(location) {
    const {
      projectPath
    } = this.options;
    const nodePosition = this.sourceMapConsumer.originalPositionFor({
      line: location.start.line,
      column: location.start.column
    });
    if (nodePosition.source?.endsWith('.ux') && nodePosition.line) {
      nodePosition.line = nodePosition.line - 1 + this.lineOffset;
    }
    return {
      pos: 0,
      end: 0,
      startLine: nodePosition.line || 0,
      startColumn: nodePosition.column || 0,
      endLine: nodePosition.line ? nodePosition.line + (location.end.line - location.start.line) : 0,
      endColumn: location.end.column - 1,
      source: nodePosition.source ? _path.default.relative(projectPath, decodeURI(nodePosition.source)) : ''
    };
  }
}
var _default = exports.default = StyleMapUtil;
