// @aiot-toolkit/parser/lib/ux/translate/vela/plugins/e2e.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = e2e;
var t = _interopRequireWildcard(require("@babel/types"));
var parser = _interopRequireWildcard(require("@babel/parser"));
var _path = _interopRequireDefault(require("path"));
var _fsExtra = _interopRequireDefault(require("fs-extra"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/**
 * 用于添加应用自动化测试代码
 * 分为两种情况
 * 1. app文件
 * 2. page 文件
 */
function e2e() {
  return {
    visitor: {
      ExportDefaultDeclaration(path, _ref) {
        let {
          opts
        } = _ref;
        const options = opts.options;
        switch (options.fileType) {
          case 'page':
            addPageTestCode(path, opts);
            break;
          case 'app':
            addAppTestCode(path);
            break;
        }
      }
    }
  };
}

/**
 * 添加 app 的测试代码
 *
 * 引入测试框架文件
 * @param path
 * @param opts
 */
function addAppTestCode(path) {
  const runtimePath = _path.default.join(__dirname, '../runtime/velaTestLibrary.js');
  const appImport = t.importDeclaration([], t.stringLiteral(runtimePath));
  if (path.parent.type === 'Program') {
    path.parent.body.unshift(appImport);
  }
}

/**
 * 添加页面的测试代码
 * 1. 检查测试文件是否存在
 * 2. 如果存在，则
 *    a. import 测试文件
 *    b. 设置测试函数: 如果 onTestStart 方法已存在，则在其内部添加测试代码；否则，添加 onTestStart方法
 *
 * @description 示例
 *
 * 源码 `src/home/myHome.ux`
 * ```
 *
 * export default {}
 * ```
 *
 * 产物
 * ```
 * import fnTestCase from 'test/home/myHome.js'
 *
 * export default {
 *  onTestStart: (){
 *      // code
 *   }
 * }
 * ```
 *
 * @param path
 * @param opts
 */
function addPageTestCode(path, opts) {
  const testFilePath = getTestFilePath(opts);
  // 1.
  if (_fsExtra.default.existsSync(testFilePath)) {
    const funcName = 'onTestStart';
    const funcBody = generateTestFn();
    // 2.a
    const vmImport = t.importDeclaration([t.importDefaultSpecifier(t.identifier('fnTestCase'))], t.stringLiteral(testFilePath));
    if (path.parent.type === 'Program') {
      path.parent.body.unshift(vmImport);
    }

    // 2.b
    if (path.node.declaration.type === 'ObjectExpression') {
      const properties = path.node.declaration.properties;
      const testFunc = properties.find(p => p.type === 'ObjectMethod' && p.key.name === funcName);
      if (testFunc) {
        testFunc.body.body.push(...funcBody.body);
      } else {
        properties.push(t.objectMethod('method', {
          name: funcName,
          type: 'Identifier'
        }, [], funcBody));
      }
    }
  }
}
function getTestFilePath(opts) {
  const {
    filePath,
    projectPath
  } = opts.options;
  const {
    sourceRoot
  } = opts.compilerOption;
  const relativeSrc = _path.default.relative(_path.default.join(projectPath, sourceRoot || ''), filePath);
  const parsed = _path.default.parse(relativeSrc);
  return _path.default.join(projectPath, 'test', parsed.dir, `${parsed.name}.js`);
}

/**
 * 获取测试函数的函数体
 * @returns
 */
function generateTestFn() {
  const fn = `function(){
  this.$options = this._options || this.$options;
  if (this.$options._descriptor) {
    this.$options._descriptor["back"] = {
      access: "public",
    };
  }
  global.CASE_TEST_START = global.CASE_TEST_START || 1000;
  global.qat = new VelaQuickappTest();
  setTimeout(() => {
    typeof fnTestCase === 'function' && fnTestCase(this)
    qat.run(() => {
      if (this.back !== "false") {
        console.info("拥有关联测试用例，测试完毕，返回到之前的页面");
        function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
        var _system = _interopRequireDefault($app_require$('@app-module/system.router'));
        _system.default.back();
      }
    });
  }, global.CASE_TEST_START);
  }`;
  return parser.parseExpression(fn).body;
}
