// ux-types/lib/componentPublicInstance.ts
import { ComputedOptions, ExtractComputedReturns, GeyMethods } from './componentOptions';
import { nextTick } from './scheduler';
import { ProjectType } from './shared';

export declare type Data = Record<string, unknown>;

export type CreateComponentPublicInstance<
	Props = {},
	Private = {},
	Protected = {},
	Public = {},
	D = {},
	C extends ComputedOptions = {},
	O = {},
	T extends ProjectType = 'quickapp'
> = ComponentPublicInstance<Props, Private, Protected, Public, D, C, O, T>;

// public properties exposed on the proxy, which is used as the render context
// in templates (as `this` in the render option)
export type ComponentPublicInstance<
	Props = {}, // props type extracted from props option
	Private = {},
	Protected = {},
	Public = {},
	D = {}, // return from data()
	C extends ComputedOptions = {},
	O = {},
	T extends ProjectType = 'quickapp'
> = GetBuildinType<T, Props> & D & Private & Protected & Public & ExtractComputedReturns<C> & GeyMethods<O> & Props;

type GetBuildinType<T extends ProjectType, Props> = T extends 'vela'
	? ComponentBuildinTypeForVela<Props>
	: T extends 'mina'
	? ComponentBuildinTypeForMina<Props>
	: ComponentBuildinType<Props>;

export interface ComponentBuildinType<Props = {}> {
	props?: Props;
	private?: any;
	protected?: any;
	public?: any;
	data?: any;
	computed?: any;

	// assets
	filters?: Record<string, Function>;

	// lifecycle
	//页面初始化
	onInit?(): void;
	onReady?(): void;
	onShow?(): void;
	onHide?(): void;
	onDestroy?(): void;
	onBackPress?(): void;
	onMenuPress?(): void;
	onRefresh?(): void;
	onConfigurationChanged?(): void;
	onReachTop?(): void;
	onReachBottom?(): void;
	onPageScroll?(event: any): void;

	$app: { $def: any; $data: any; exit: () => void };
	$page: {
		action: string;
		uri: string;
		name: string;
		path: string;
		component: string;
		orientation: string;
		statusBarHeight: number;
		titleBarHeight: number;
		windowWidth: number;
		windowHeight: number;
		setTitleBar: (
			options: Partial<{
				text: string;
				textColor: string;
				backgroundColor: string;
				backgroundOpacity: number;
				menu: boolean;
			}>
		) => void;
		setMeta: (options: Partial<{ title: string; description: string; pic: string; keywords: string }>) => void;
		finish: () => void;
		exitFullscreen: () => void;
		setStatusBar: (
			options: Partial<{ immersive: boolean; textStyle: string; backgroundColor: string; backgroundOpacity: number }>
		) => void;
		scrollTo: (options: Partial<{ top: number; behavior: 'smooth' | 'instant' | 'auto' }>) => void;
		scrollBy: (options: Partial<{ top: number; behavior: 'smooth' | 'instant' | 'auto' }>) => void;
		getMenuBarRect: () => any;
		getMenuBarBoundingRect: () => any;
		setMenubarData: (
			options: Partial<{
				shareTitle: string;
				shareDescription: string;
				shareIcon: string;
				shareCurrentPage: boolean;
				shareParams: any;
				shareUrl: string;
				usePageParams: boolean;
			}>
		) => void;
		setSecure: (isSecure: boolean) => void;
	};
	$vaild: boolean;
	$visible: boolean;
	$attrs: Data;
	$listeners: any;

	$set: (key: string, value: any) => void;
	$delete: (key: string) => void;

	$element: (id: string) => any;
	$root: () => void;
	$parnet: () => void;
	$child: () => void;
	$nextTick: typeof nextTick;
	$forceUpdate: () => void;

	/** @deprecated use `$child` instead */
	$vm: (id: string) => any;
	/** @deprecated use `$element` instead */
	$rootElement: () => any;

	$on: (type: string, handler: () => any) => void;
	$off: (type: string, handler?: () => any) => void;
	$dispatch: (type: string) => void;
	$broadcast: (type: string) => void;
	$emit: (type: string, data: any) => void;
	$emitElement: (type: string, data: any, id: string) => void;
	$watch: (data: string, handler: string) => void;

	$t: (path: string, arg0?: any) => string;
	$tc: (path: string, count: number) => string;
}

export interface ComponentBuildinTypeForVela<Props = {}> {
	props?: Props;
	private?: any;
	protected?: any;
	public?: any;
	data?: any;
	computed?: any;

	// lifecycle
	//页面初始化
	onInit?(): void;
	onReady?(): void;
	onShow?(): void;
	onHide?(): void;
	onDestroy?(): void;
	onBackPress?(): void;

	$app: { $def: any; $data: any };
	$page: {
		action: string;
		uri: string;
		name: string;
		path: string;
		component: string;
		orientation: string;
		statusBarHeight: number;
		titleBarHeight: number;
		windowWidth: number;
		windowHeight: number;
	};

	$element: (id: string) => any;

	$t: (path: string, arg0?: any) => string;
	$tc: (path: string, count: number) => string;
}

export interface ComponentBuildinTypeForMina<Props = {}> {
	props?: Props;
	private?: any;
	protected?: any;
	public?: any;
	data?: any;
	computed?: any;

	// 生命周期函数
	onInit?(): void;
	onReady?(): void;
	onShow?(): void;
	onHide?(): void;
	onDestroy?(): void;
	onBackPress?(): void;
	onMenuPress?(): void;
	onRefresh?(): void;
	onConfigurationChanged?(): void;

	//公共对象
	$app: { $def: any; $data: any; exit: () => void };
	$page: {
		name: string;
		path: string;
		component: string;
		VIEW_ORIENTATION: string;
		query: any;
		// 仅定义，未实现
		setTitleBar: (
			options: Partial<{
				text: string;
				textColor: string;
				backgroundColor: string;
				backgroundOpacity: number;
				menu: boolean;
			}>
		) => void;
		setMeta: (options: Partial<{ title: string; description: string; pic: string; keywords: string }>) => void;
		finish: () => void;
		exitFullscreen: () => void;
		setStatusBar: (
			options: Partial<{ immersive: boolean; textStyle: string; backgroundColor: string; backgroundOpacity: number }>
		) => void;
		scrollTo: (options: Partial<{ top: number; behavior: 'smooth' | 'instant' | 'auto' }>) => void;
		scrollBy: (options: Partial<{ top: number; behavior: 'smooth' | 'instant' | 'auto' }>) => void;
		getMenuBarRect: () => any;
		setMenubarData: (options: any) => void;
		// 实现的方法
		setPageStyle: (options: Partial<{ backgroundColor: string }>) => void;
	};
	$visible: boolean;

	//公共方法
	$set: (key: string, value: any) => void;
	$delete: (key: string) => void;

	$element: (id: string) => any;
	$root: () => void;
	$parent: () => void;
	$child: () => void;
	$nextTick: typeof nextTick;
	$forceUpdate: () => void;

	$on: (type: string, handler: () => any) => void;
	$off: (type: string, handler?: () => any) => void;
	$dispatch: (type: string) => void;
	$broadcast: (type: string) => void;
	$emit: (type: string, data: any) => void;
	$emitElement: (type: string, data: any, id: string) => void;
	$watch: (data: string, handler: string) => void;

	$t: (path: string, arg0?: any) => string;
	$tc: (path: string, count: number) => string;
}
